### PR TITLE
Communication server implementation.

### PR DESCRIPTION
**Pull request by codito**
Extract socket implementation to allow experimentation with multiple
data interchange formats and ipc.
Add concept of framing for messages passed between various processes.
TestRequestSender2 uses these concepts and is a replacement for the
earlier TestRequestSender.

### CODE REVIEW

File: test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        public async Task SocketServerStartShouldHostServer()
+        {
+            var connectionInfo = this.socketServer.Start();
+
+            Assert.IsFalse(string.IsNullOrEmpty(connectionInfo));
+            await this.ConnectToServer(int.Parse(connectionInfo));
+            Assert.IsTrue(this.tcpClient.Connected);
+        }
+
+        ////[TestMethod]
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CommunicationUtilities.PlatformTests/SocketServerTests.cs]: Why has this been commented out ?

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        private ICommunicationChannel channel;
+
+        private TcpListener tcpListener;
+
+        private TcpClient tcpClient;
+
+        private bool stopped;
+
+        ////private CancellationTokenSource cancellationToken;
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs]: Suggestion : Can be removed.

File: src/Microsoft.TestPlatform.CommunicationUtilities/Microsoft.TestPlatform.CommunicationUtilities.csproj [hunk 3]
```diff
@@ -18,19 +18,21 @@
...
+    <PackageReference Include="Newtonsoft.Json">
+      <Version>9.0.1</Version>
+    </PackageReference>
   </ItemGroup>
   <ItemGroup Condition=" '$(TargetFramework)' == 'net46' ">
     <Reference Include="System" />
     <Reference Include="Microsoft.CSharp" />
+
+    <PackageReference Include="Newtonsoft.Json">
+      <Version>8.0.3</Version>
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/Microsoft.TestPlatform.CommunicationUtilities.csproj]: Yes, it is intentional. `8.0.3` does not support netcoreapp scenarios.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                EqtTrace.Info("SocketClient: Stop: Cancellation requested. Stopping message loop.");
+                this.cancellation.Cancel();
+            }
+        }
+
+        private void OnServerConnected(Task connectAsyncTask)
+        {
+            if (connectAsyncTask.IsFaulted)
+            {
+                // Throw an exception
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Throw an exception.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            }
+
+            this.channel = this.channelFactory(this.tcpClient.GetStream());
+            if (this.ServerConnected != null)
+            {
+                this.ServerConnected.Invoke(this, new ConnectedEventArgs(this.channel));
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Suggestion : Use [SafeInvoke](https://github.com/Microsoft/vstest/blob/c701070abb126853b58800cedd7554482471ac7d/src/Microsoft.TestPlatform.CoreUtilities/Utilities/MulticastDelegateUtilities.cs#L26) as it takes care of EqtTrace logging as well.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                // Stop accepting any other connections
+                this.tcpListener.Stop();
+
+                // Close the client and dispose the underlying stream
+                this.tcpClient?.Dispose();
+                this.channel.Dispose();
+
+                this.cancellation.Dispose();
+
+                if (this.ClientDisconnected != null)
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketServer.cs]: Can be replaced with one line `this.ClientDisconnected?.Invoke`

File: src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender2.cs [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+        public void EndSession()
+        {
+            if (!this.IsOperationComplete())
+            {
+                this.channel.Send(this.dataSerializer.SerializeMessage(MessageType.SessionEnd));
+            }
+        }
+
+        /// <inheritdoc />
+        public void OnClientProcessExit(string stdError)
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender2.cs]: Is it used anywhere?

Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/TestRequestSender2.cs]: Implementation of `ITestRequestSender.OnClientProcessExit`, used to notify if test host has exited by callers.

**Pull request merged**